File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
class Solution {
    static int[][] dp;
    public int solution(int[][] triangle) {
        dp=new int[triangle.length+1][triangle.length+1];
        
        for (int i=1; i<=triangle.length; i++){
            for (int j=1; j<=i; j++){
                dp[i][j]=Math.max(dp[i-1][j-1], dp[i-1][j])+triangle[i-1][j-1];
            }
        }
        
        int answer=dp[triangle.length][0];
        for (int i=0; i<triangle.length+1; i++){
            answer=Math.max(answer, dp[triangle.length][i]);
        }
        
        return answer;
    }
}

File: C++17/백준/Silver/1932. 정수 삼각형/정수 삼각형.cc
#include <iostream>
#include <climits>
#include <cstring>
using namespace std;

int main(){
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int n; cin >> n;

    int triangle[n][n];
    for (int i = 0; i < n; i++){
        for (int j = 0; j <= i; j++){
            int x; cin >> x;

            triangle[i][j] = x;
        }
    }

    int dp[n][n];
    memset(dp, 0, sizeof(dp));
    dp[0][0] = triangle[0][0];
    for (int i = 1; i < n; i++){
        for (int j = 0; j <= i; j++){
            dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - 1]) + triangle[i][j];
        }
    }

    int max_res = INT_MIN;
    for (int i = 0; i < n; i++){
        max_res = max(max_res, dp[n - 1][i]);
    }

    cout << max_res << '\n';

    return 0;
}